crates/swc_cli/src/commands/compile.rs

```diff
@@ -1,5 +1,220 @@
+
+    /// Files to compile
+    files: Vec<PathBuf>,
+
+    /// The output directory
+    #[clap(long)]
+    out_dir: Option<PathBuf>,
+
+    /// Specify specific file extensions to compile.
+    #[clap(long)]
+    extensions: Option<Vec<String>>,
+}
+
+static COMPILER: Lazy<Arc<Compiler>> = Lazy::new(|| {
+    let cm = Arc::new(SourceMap::new(FilePathMapping::empty()));
+
+    Arc::new(Compiler::new(cm))
+});
+
+/// List of file extensions supported by default.
+static DEFAULT_EXTENSIONS: &[&str] = &["js", "jsx", "es6", "es", "mjs", "ts", "tsx"];
+
+/// Infer list of files to be transformed from cli arguments.
+/// If given input is a directory, it'll traverse it and collect all supported
+/// files.
+fn get_files_list(
+    raw_files_input: &[PathBuf],
+    extensions: &[String],
+    _include_dotfiles: bool,
+) -> anyhow::Result<Vec<PathBuf>> {
+    let input_dir = raw_files_input.iter().find(|p| p.is_dir());
+
+    let files = if let Some(input_dir) = input_dir {
+        if raw_files_input.len() > 1 {
+            return Err(anyhow::anyhow!(
+                "Cannot specify multiple files when using a directory as input"
+            ));
+        }
+
+        WalkDir::new(input_dir)
+            .into_iter()
+            .filter_map(|e| e.ok())
+            .map(|e| e.into_path())
+            .filter(|e| {
+                extensions.iter().any(|ext| {
+                    e.extension()
+                        .map(|v| v.to_str().unwrap_or("").eq(ext))
+                        .unwrap_or(false)
+                })
+            })
+            .collect()
+    } else {
+        raw_files_input.to_owned()
+    };
+
+    Ok(files)
+}
+
+fn build_transform_options(
+    config_file: &Option<PathBuf>,
+    file_path: &Path,
+) -> anyhow::Result<Options> {
+    let base_options = Options::default();
+    let base_config = Config::default();
+
+    let config_file = if let Some(config_file_path) = config_file {
+        let config_file_contents = fs::read(config_file_path)?;
+        serde_json::from_slice(&config_file_contents)
+            .map_err(|e| anyhow::anyhow!("{}", e))
+            .context("Failed to parse config file")?
+    } else {
+        None
+    };
+
+    Ok(Options {
+        config: Config { ..base_config },
+        config_file,
+        filename: file_path.to_str().unwrap_or_default().to_owned(),
+        ..base_options
+    })
+}
+
+/// Calculate full, absolute path to the file to emit.
+/// Currently this is quite naive calculation based on assumption input file's
+/// path and output dir are relative to the same directory.
+fn resolve_output_file_path(out_dir: &Path, file_path: &Path) -> anyhow::Result<PathBuf> {
+    let default = PathBuf::from(".");
+    let base = file_path
+        .parent()
+        .unwrap_or(&default)
+        .as_os_str()
+        .to_string_lossy();
+
+    let dist_absolute_path = out_dir.absolutize()?;
+
+    // These are possible combinations between input to output dir.
```

Comment:
This is note for the future what we'll need to support for the path calculation.